src/overlaybd/tar/erofs/liberofs.cpp

```diff
@@ -0,0 +1,661 @@
+}
+
+static off_t erofs_target_lseek(struct erofs_vfile *vf, u64 offset, int whence)
+{
+    return -EROFS_UNIMPLEMENTED;
+}
+
+/* I/O control for source */
+static ssize_t erofs_source_pread(struct erofs_vfile *vf, void *buf, u64 offset,
+                                  size_t len)
+{
+    return -EROFS_UNIMPLEMENTED;
+}
+
+static ssize_t erofs_source_pwrite(struct erofs_vfile *vf, const void *buf,
+                                   u64 offset, size_t len)
+{
+    return -EROFS_UNIMPLEMENTED;
+}
+
+static int erofs_source_fsync(struct erofs_vfile *vf)
+{
+    return -EROFS_UNIMPLEMENTED;
+}
+
+static int erofs_source_fallocate(struct erofs_vfile *vf,
+                                  u64 offset, size_t len, bool pad)
+{
+    return -EROFS_UNIMPLEMENTED;
+}
+
+static int erofs_source_ftruncate(struct erofs_vfile *vf, u64 length)
+{
+    return EROFS_UNIMPLEMENTED;
+}
+
+static ssize_t erofs_source_read(struct erofs_vfile *vf, void *buf,
+                                 size_t bytes)
+{
+    u64 i = 0;
+    while (bytes) {
+        u64 len = bytes > INT_MAX ? INT_MAX : bytes;
+        u64 ret;
+
+        ret = _source->read(buf + i, len);
+        if (ret < 1) {
+            if (ret == 0)
+                break;
+            else
+                return -1;
+        }
+        bytes -= ret;
+        i += ret;
+    }
+    return i;
+}
+
+static off_t erofs_source_lseek(struct erofs_vfile *vf, u64 offset, int whence)
+{
+    return _source->lseek(offset, whence);
+}
+
+struct erofs_mkfs_cfg {
+    struct erofs_sb_info *sbi;
+    struct erofs_tarfile *erofstar;
+    bool incremental;
+    bool ovlfs_strip;
+};
+
+static int rebuild_src_count;
+
+int erofs_mkfs(struct erofs_mkfs_cfg *cfg)
+{
+    int err;
+    struct erofs_tarfile *erofstar;
+    struct erofs_sb_info *sbi;
+    struct erofs_buffer_head *sb_bh;
+    struct erofs_inode *root;
+    erofs_blk_t nblocks;
+
+    erofstar = cfg->erofstar;
+    sbi = cfg->sbi;
+    if (!erofstar || !sbi)
+        return -EINVAL;
+
+    if (!erofstar->mapfile)
+        return -EINVAL;
+
+    err = erofs_blocklist_open(erofstar->mapfile, true);
+    if (err) {
+        LOG_ERROR("[erofs] Fail to open erofs blocklist.");
+        return -EINVAL;
+    }
+
+    if (!erofstar->rvsp_mode) {
+        LOG_ERROR("[erofs] Must be in RVSP mode.");
+        return -EINVAL;
+    }
+
+    if (!cfg->incremental) {
+        sb_bh = erofs_reserve_sb(sbi);
+        if (IS_ERR(sb_bh)) {
+            LOG_ERROR("[erofs] Fail to reseve space for superblock.");
+            err =  PTR_ERR(sb_bh);
+            goto exit;
+        }
+        //erofs_uuid_generate(sbi->uuid);
+    } else {
+        err = erofs_read_superblock(sbi);
+        if (err) {
+            LOG_ERROR("[erofs] Fail to read superblock.");
+            goto exit;
+        }
+        erofs_buffer_init(sbi, sbi->primarydevice_blocks);
+        sb_bh = NULL;
+    }
+
+    erofs_inode_manager_init();
+
+    root = erofs_mkfs_alloc_root(sbi);
```

Comment:
let's rename this as erofs_rebuild_make_root(sbi);

Reply:
Done.